dhall-csv/src/Dhall/CsvToDhall.hs

```diff
@@ -1,27 +1,254 @@
+        = Left $ UnicodeError (head badKeys) -- Only report first key that failed to be decoded
+        | extraKeys <- (map decodeUtf8 $ HashMap.keys csvRecord) \\ Map.keys record
+        , strictRecs && not (null extraKeys)
+        = Left $ UnhandledKeys extraKeys
+        | otherwise
+        = do
+            let f k v = fieldConvert k (Core.recordFieldValue v) (HashMap.lookup (encodeUtf8 k) csvRecord)
+            a <- Map.traverseWithKey (\k v -> mapRight Core.makeRecordField (f k v)) record
+            return $ Core.RecordLit a
+    recordConvert e _ = Left $ Unsupported e
+
+    fieldConvert :: Text -> ExprX -> Maybe Data.Csv.Field -> Either CompileError ExprX
+    -- Unions
+    fieldConvert recordKey t@(Core.Union tm) maybeField = do
+        let f unionKey Nothing =
+                case maybeField of
+                    Nothing -> Left $ MissingKey recordKey
+                    Just field ->
+                        case decodeUtf8' field of
+                            Left err -> Left $ UnicodeError err
+                            Right _field ->
+                                if _field == unionKey
+                                then Right $ Core.Field t $ Core.makeFieldSelection unionKey
+                                else Left $ Mismatch t field recordKey
+            f unionKey (Just _type) = do
+                expression <- fieldConvert recordKey _type maybeField
+                return (Core.App (Core.Field t $ Core.makeFieldSelection unionKey) expression)
+
+        case (unions, rights (toList (Map.mapWithKey f tm)), maybeField) of
+            (UNone  , _         , _         ) -> Left $ ContainsUnion t
+            (UStrict, xs@(_:_:_), Just field) -> Left $ UndecidableUnion t field xs
+            (UStrict, xs@(_:_:_), Nothing   ) -> Left $ UndecidableMissingUnion t xs
+            (_      , []        , Just field) -> Left $ Mismatch t field recordKey
+            (_      , []        , Nothing   ) -> Left $ MissingKey recordKey
+            (UFirst , x:_       , _         ) -> Right $ x
+            (UStrict, [x]       , _         ) -> Right $ x
+
+    -- Missing Optionals
+    fieldConvert _ (Core.App Core.Optional t) Nothing = return $ Core.App Core.None t
+
+    -- Missing fields
+    fieldConvert key _ Nothing = Left $ MissingKey key
+
+    -- Bools
+    fieldConvert key Core.Bool (Just field) =
+        case field of
+            "true"  -> Right (Core.BoolLit True)
+            "false" -> Right (Core.BoolLit False)
+            _ -> Left $ Mismatch Core.Bool field key
+
+    -- Naturals
+    fieldConvert key Core.Natural (Just field) =
+        case decodeUtf8' field of
+            Left err -> Left $ UnicodeError err
+            Right _field ->
+                case decimal _field of
+                    Left _ -> Left $ Mismatch Core.Natural field key
+                    Right (v, _) -> Right $ Core.NaturalLit v           -- What to do when there is more text left to read?
```

Comment:
This is how you can easily enforce that there is no trailing `Text`:

```suggestion
                    Right (v, "") -> Right $ Core.NaturalLit v           -- What to do when there is more text left to read?
                    _ -> Left $ Mismatch Core.Natural field key
```

Same trick for the other numeric types

Reply:
I thought of that but wasn't sure of the case when there are trailing whitespaces. I suppose we can forbid any trailing text for now and later think of something to handle that case

Reply:
I think it's fine if we're strict about not allowing leading/trailing whitespace

Reply:
Great! I'll fix it and merge the branch